File: knowledgebase-sdk/src/main/java/ru/usedesk/knowledgebase_sdk/external/UsedeskKnowledgeBaseSdk.java
package ru.usedesk.knowledgebase_sdk.external;

import android.content.Context;

import androidx.annotation.NonNull;

import ru.usedesk.knowledgebase_sdk.external.entity.UsedeskKnowledgeBaseConfiguration;
import ru.usedesk.knowledgebase_sdk.internal.di.InstanceBox;

public final class UsedeskKnowledgeBaseSdk {

    private static InstanceBox instanceBox;
    private static UsedeskKnowledgeBaseConfiguration configuration;

    @NonNull
    public static IUsedeskKnowledgeBase init(@NonNull Context appContext) {
        if (instanceBox == null) {
            if (configuration == null) {
                throw new RuntimeException("Must call UsedeskKnowledgeBaseSdk.setConfiguration(...) before");
            }
            instanceBox = new InstanceBox(appContext, configuration);
        }
        return instanceBox.getKnowledgeBaseSdk();
    }

    @NonNull
    public static IUsedeskKnowledgeBase getInstance() {
        if (instanceBox == null) {
            throw new RuntimeException("Must call UsedeskKnowledgeBaseSdk.init(...) before");
        }
        return instanceBox.getKnowledgeBaseSdk();
    }

    public static void setConfiguration(@NonNull UsedeskKnowledgeBaseConfiguration knowledgeBaseConfiguration) {
        configuration = knowledgeBaseConfiguration;
    }

    public static void release() {
        if (instanceBox != null) {
            instanceBox.release();
            instanceBox = null;
        }
    }
}

File: knowledgebase-sdk/src/main/java/ru/usedesk/knowledgebase_sdk/external/UsedeskKnowledgeBaseSdk.kt
package ru.usedesk.knowledgebase_sdk.external

import android.content.Context
import ru.usedesk.knowledgebase_sdk.external.entity.UsedeskKnowledgeBaseConfiguration
import ru.usedesk.knowledgebase_sdk.internal.di.InstanceBox

object UsedeskKnowledgeBaseSdk {
    private var instanceBox: InstanceBox? = null
    private var configuration: UsedeskKnowledgeBaseConfiguration? = null

    @JvmStatic
    fun init(appContext: Context): IUsedeskKnowledgeBase {
        if (instanceBox == null) {
            checkConfiguration()
            instanceBox = InstanceBox(appContext, configuration!!)
        }
        return instanceBox!!.knowledgeBaseSdk
    }

    @JvmStatic
    fun getInstance(): IUsedeskKnowledgeBase {
        if (instanceBox == null) {
            throw RuntimeException("Must call UsedeskKnowledgeBaseSdk.init(...) before")
        }
        return instanceBox!!.knowledgeBaseSdk
    }

    @JvmStatic
    fun setConfiguration(knowledgeBaseConfiguration: UsedeskKnowledgeBaseConfiguration) {
        configuration = knowledgeBaseConfiguration
    }

    @JvmStatic
    fun release() {
        if (instanceBox != null) {
            instanceBox!!.release()
            instanceBox = null
        }
    }

    private fun checkConfiguration() {
        if (configuration == null) {
            throw RuntimeException("Must call UsedeskKnowledgeBaseSdk.setConfiguration(...) before")
        }
    }
}